File: Python/currency_converter.py
from typing import Dict, Optional
import threading
import time
import datetime
import json
import urllib.request


class CurrencyConverter:
    _instance = None
    _lock = threading.Lock()

    def __new__(cls):
        with cls._lock:
            if cls._instance is None:
                cls._instance = super().__new__(cls)
                cls._instance._rates = {}
                cls._instance._last_update = 0
                cls._instance._updating = False
                cls._instance._rate_times = {}
            return cls._instance

    def __init__(self):
        pass

    def get_rate(self, from_currency: str, to_currency: str) -> float:
        if from_currency == to_currency:
            return 1.0

        key = (from_currency, to_currency)
        if key in self._rates:
            return self._rates[key]

        rate = self._fetch_rate(from_currency, to_currency)
        if rate is None:
            rate = 1.0
        self._rates[key] = rate
        self._rate_times[key] = datetime.datetime.now()
        return rate

    def get_rate_time(self, from_currency: str, to_currency: str):
        return self._rate_times.get((from_currency, to_currency))

    def get_all_rates(self, base_currency: str, currencies):
        rows = []
        for code in currencies:
            if code == base_currency:
                rows.append((code, 1.0, None))
                continue
            try:
                rate = self.get_rate(base_currency, code)
            except Exception:
                rate = 1.0
            t = self._rate_times.get((base_currency, code))
            rows.append((code, rate, t))
        return rows

    def _fetch_rate(self, from_currency: str, to_currency: str) -> Optional[float]:
        # Live rates from open.er-api.com (free, no API key required), same source
        # the Android app uses, so both platforms compute identical conversions.
        try:
            url = f"https://open.er-api.com/v6/latest/{from_currency}"
            req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
            with urllib.request.urlopen(req, timeout=8) as resp:
                data = json.loads(resp.read().decode("utf-8"))
            if data.get("result") == "success" and "rates" in data:
                rates = data["rates"]
                if to_currency in rates:
                    rate = float(rates[to_currency])
                    self._last_update = time.time()
                    return rate
        except Exception as e:
            print(f"Failed to fetch exchange rate: {e}")

        fallback_rates = {
            ("HKD", "CNY"): 0.92,
            ("CNY", "HKD"): 1.087,
            ("USD", "CNY"): 7.25,
            ("CNY", "USD"): 0.138,
            ("USD", "HKD"): 7.82,
            ("HKD", "USD"): 0.128,
            ("EUR", "CNY"): 7.85,
            ("CNY", "EUR"): 0.127,
            ("GBP", "CNY"): 9.15,
            ("CNY", "GBP"): 0.109,
            ("JPY", "CNY"): 0.048,
            ("CNY", "JPY"): 20.83,
        }

        key = (from_currency, to_currency)
        if key in fallback_rates:
            return fallback_rates[key]

        reverse_key = (to_currency, from_currency)
        if reverse_key in fallback_rates:
            return 1.0 / fallback_rates[reverse_key]

        return 1.0

    def convert(self, amount: float, from_currency: str, to_currency: str) -> float:
        rate = self.get_rate(from_currency, to_currency)
        return amount * rate

    def refresh_rates(self):
        self._rates.clear()
        self._rate_times.clear()
        self._last_update = 0

    def prefetch_rates(self, currencies, base_currency: str):
        """Fetch rates only for the given currencies against the base currency."""
        for cur in set(currencies):
            if cur != base_currency:
                try:
                    self.get_rate(cur, base_currency)
                except Exception:
                    pass

    def get_supported_currencies(self) -> list:
        return ["CNY", "HKD", "USD", "EUR", "GBP", "JPY", "TWD", "KRW", "SGD", "AUD", "CAD"]

    def validate_currency(self, code: str) -> bool:
        """Return True if `code` looks like a real currency.

        Checks the code against the live open.er-api.com symbol list; if the
        network is unavailable it falls back to accepting any well-formed
        3-letter alphabetic code.
        """
        code = (code or "").strip().upper()
        if len(code) != 3 or not code.isalpha():
            return False
        try:
            url = "https://open.er-api.com/v6/latest/USD"
            req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
            with urllib.request.urlopen(req, timeout=8) as resp:
                data = json.loads(resp.read().decode("utf-8"))
            rates = data.get("rates", {})
            return code in rates
        except Exception:
            return True


def get_used_currencies(db) -> list:
    """Collect all currencies actually used by accounts and debts."""
    currencies = {db.get_base_currency()}
    for acc in db.get_accounts():
        currencies.add(acc.currency)
    for debt in db.get_debts():
        currencies.add(debt.currency)
    return sorted(currencies)

File: Android/app/src/main/java/com/example/sync/CurrencyConverter.kt
package com.example.sync

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

object CurrencyConverter {
    private val ratesCache = ConcurrentHashMap<Pair<String, String>, Double>()
    private val rateTimes = ConcurrentHashMap<Pair<String, String>, Long>()
    private val client = OkHttpClient.Builder()
        .connectTimeout(8, TimeUnit.SECONDS)
        .readTimeout(8, TimeUnit.SECONDS)
        .build()

    val supportedCurrencies = listOf(
        "CNY", "HKD", "USD", "EUR", "GBP", "JPY", "TWD", "KRW", "SGD", "AUD", "CAD"
    )

    private val fallbackRates = mapOf(
        Pair("HKD", "CNY") to 0.92,
        Pair("CNY", "HKD") to 1.087,
        Pair("USD", "CNY") to 7.25,
        Pair("CNY", "USD") to 0.138,
        Pair("USD", "HKD") to 7.82,
        Pair("HKD", "USD") to 0.128,
        Pair("EUR", "CNY") to 7.85,
        Pair("CNY", "EUR") to 0.127,
        Pair("GBP", "CNY") to 9.15,
        Pair("CNY", "GBP") to 0.109,
        Pair("JPY", "CNY") to 0.048,
        Pair("CNY", "JPY") to 20.83,
        Pair("USD", "EUR") to 0.92,
        Pair("EUR", "USD") to 1.087,
        Pair("USD", "GBP") to 0.79,
        Pair("GBP", "USD") to 1.265,
        Pair("USD", "JPY") to 155.0,
        Pair("JPY", "USD") to 0.00645,
        Pair("USD", "TWD") to 32.5,
        Pair("TWD", "USD") to 0.0308,
        Pair("USD", "KRW") to 1380.0,
        Pair("KRW", "USD") to 0.00072,
        Pair("USD", "SGD") to 1.35,
        Pair("SGD", "USD") to 0.74,
        Pair("USD", "AUD") to 1.52,
        Pair("AUD", "USD") to 0.658,
        Pair("USD", "CAD") to 1.36,
        Pair("CAD", "USD") to 0.735
    )

    fun symbol(currency: String): String {
        return when (currency.uppercase()) {
            "CNY" -> "¥"
            "HKD" -> "HK$"
            "USD" -> "$"
            "EUR" -> "€"
            "GBP" -> "£"
            "JPY" -> "¥"
            "TWD" -> "NT$"
            "KRW" -> "₩"
            "SGD" -> "S$"
            "AUD" -> "A$"
            "CAD" -> "C$"
            else -> "$currency "
        }
    }

    fun getRate(fromCurrency: String, toCurrency: String): Double {
        val from = fromCurrency.uppercase().trim()
        val to = toCurrency.uppercase().trim()
        if (from == to || from.isEmpty() || to.isEmpty()) return 1.0

        val key = Pair(from, to)
        ratesCache[key]?.let { return it }

        fallbackRates[key]?.let {
            ratesCache[key] = it
            return it
        }

        val reverseKey = Pair(to, from)
        fallbackRates[reverseKey]?.let {
            val inverted = 1.0 / it
            ratesCache[key] = inverted
            return inverted
        }

        // Cross-rate via USD
        val fromToUsd = if (from == "USD") 1.0 else (fallbackRates[Pair(from, "USD")] ?: (1.0 / (fallbackRates[Pair("USD", from)] ?: 1.0)))
        val usdToTarget = if (to == "USD") 1.0 else (fallbackRates[Pair("USD", to)] ?: (1.0 / (fallbackRates[Pair(to, "USD")] ?: 1.0)))
        val rate = fromToUsd * usdToTarget
        ratesCache[key] = rate
        return rate
    }

    fun convert(amount: Double, fromCurrency: String, toCurrency: String): Double {
        return amount * getRate(fromCurrency, toCurrency)
    }

    fun getRateTime(fromCurrency: String, toCurrency: String): Long? {
        return rateTimes[Pair(fromCurrency.uppercase(), toCurrency.uppercase())]
    }

    fun refreshCache() {
        ratesCache.clear()
        rateTimes.clear()
    }

    suspend fun fetchLiveRates(baseCurrency: String, currencies: List<String>): Map<String, Double> {
        return withContext(Dispatchers.IO) {
            val base = baseCurrency.uppercase()
            val targets = currencies.map { it.uppercase() }.distinct()
            val result = mutableMapOf<String, Double>()
            result[base] = 1.0
            val now = System.currentTimeMillis()

            // base -> cur：直接抓 latest/{base} 讀 rates[cur]
            // （與 Python 端 get_rate(base_currency, code)/get_all_rates() 一致）
            try {
                val request = Request.Builder().url("https://open.er-api.com/v6/latest/$base").build()
                val response = client.newCall(request).execute()
                if (response.isSuccessful) {
                    val body = response.body?.string()
                    if (body != null) {
                        val json = JSONObject(body)
                        if (json.optString("result") == "success" && json.has("rates")) {
                            val ratesObj = json.getJSONObject("rates")
                            for (c in targets) {
                                if (c == base) continue
                                if (ratesObj.has(c)) {
                                    val r = ratesObj.getDouble(c) // base -> cur 直接匯率
                                    result[c] = r
                                    ratesCache[Pair(base, c)] = r
                                    rateTimes[Pair(base, c)] = now
                                }
                            }
                        }
                    }
                }
            } catch (e: Exception) {
                // 忽略：單一請求失敗不影響其餘幣別
            }

            // cur -> base：對每個幣別抓 latest/{cur} 直接讀 rates[base]
            // （與 Python 端 get_rate(cur, base_currency) 一致；不從 base 表取倒數，
            //   避免倒數捨入差——CNY 大部位會差約 0.81 HKD）
            for (cur in targets) {
                val c = cur.uppercase()
                if (c == base) continue
                try {
                    val request = Request.Builder().url("https://open.er-api.com/v6/latest/$c").build()
                    val response = client.newCall(request).execute()
                    if (response.isSuccessful) {
                        val body = response.body?.string()
                        if (body != null) {
                            val json = JSONObject(body)
                            if (json.optString("result") == "success" && json.has("rates")) {
                                val ratesObj = json.getJSONObject("rates")
                                if (ratesObj.has(base)) {
                                    val direct = ratesObj.getDouble(base) // cur -> base 直接匯率
                                    ratesCache[Pair(c, base)] = direct
                                    rateTimes[Pair(c, base)] = now
                                }
                            }
                        }
                    }
                } catch (e: Exception) {
                    // 單一幣別失敗：退化為離線/靜態匯率
                    if (!ratesCache.containsKey(Pair(c, base))) {
                        ratesCache[Pair(c, base)] = getRate(c, base)
                        ratesCache[Pair(base, c)] = getRate(base, c)
                    }
                }
            }
            result
        }
    }
}
